include/swift/RemoteInspection/TypeRefBuilder.h

```diff
@@ -796,6 +799,296 @@ class TypeRefBuilder {
       }
     }
 
+    template <template <typename Runtime> class ObjCInteropKind,
+              unsigned PointerSize>
+    void dumpTypeMetadataSection(std::ostream &stream) {
+      using Runtime = External<RuntimeTarget<PointerSize>>;
+      using Record = TargetTypeMetadataRecord<Runtime>;
+
+      // Sometimes the compiler emits multiple records pointing to the same
+      // descriptor. Track what we've seen in a set and don't dump the same
+      // thing more than once.
+      std::unordered_set<uint64_t> dumpedDescriptors;
+
+      for (const auto &reflectionInfo : ReflectionInfos) {
+        for (const auto &section :
+             {reflectionInfo.TypeMetadata, reflectionInfo.TypeMetadata2}) {
+          if (section.size() == 0)
+            break;
+
+          auto begin = section.startAddress();
+          auto end = section.endAddress();
+
+          auto size = end.getAddressData() - begin.getAddressData();
+          if (size % sizeof(Record) != 0) {
+            stream << "Typeref section size " << size
+                   << " is not evenly divisible by record size "
+                   << sizeof(Record) << " - section is invalid.\n";
+            break;
+          }
+
+          for (auto addr = begin; addr != end;
+               addr = addr.atByteOffset(sizeof(Record))) {
+            auto index = (addr.getAddressData() - begin.getAddressData()) /
+                         sizeof(Record);
+            auto *record =
+                reinterpret_cast<const Record *>(addr.getLocalBuffer());
+
+            // The record may be a direct or indirect relative pointer. Resolve
+            // the address from the record.
+            remote::RemoteAddress descriptorAddress{nullptr};
+            if (auto *directPtr = record->getDirectPointer()) {
+              descriptorAddress = remote::RemoteAddress(
+                  addr.atByteOffset(directPtr->getOffset()).getAddressData());
+            } else if (auto indirectPtr = record->getIndirectPointer()) {
+              // We won't try to handle indirect pointers. They'll point to
+              // descriptors in other images and we don't care to dump those.
+              // They'll be included in that image's types if we dump that
+              // image's types directly.
+              continue;
```

Comment:
I wonder if it might make sense to generate *some* output for this case, just to indicate that there's an indirect type descriptor here? Obviously we can't dump the actual descriptor because we won't have it, but maybe we could get the mangled name from the symbol in the relocation?

Reply:
I think it would be better not to. We're not really dumping these sections, we're dumping the types and the list of types in these sections is just how we find them. If we also want to dump the indirect pointers, I'd say emit a separate list in the output, and do it as a separate task.